File: packages/domain-floorplan/src/wall.ts
import { DomainError } from "./domain-error.js";
import type { WallKind } from "./kinds.js";
import { WALL_KINDS } from "./kinds.js";
import { LengthMm } from "./length-mm.js";
import { type Opening } from "./opening.js";
import { type PointMm } from "./point-mm.js";

export interface WallProps {
  id: string;
  start: PointMm;
  end: PointMm;
  thickness: LengthMm;
  height: LengthMm;
  kind: WallKind;
  openings?: readonly Opening[];
}

export class Wall {
  readonly id: string;
  readonly start: PointMm;
  readonly end: PointMm;
  readonly thickness: LengthMm;
  readonly height: LengthMm;
  readonly kind: WallKind;
  readonly openings: readonly Opening[];
  readonly lengthMm: LengthMm;

  private constructor(props: WallProps, openings: readonly Opening[]) {
    this.id = props.id;
    this.start = props.start;
    this.end = props.end;
    this.thickness = props.thickness;
    this.height = props.height;
    this.kind = props.kind;
    this.openings = openings;
    this.lengthMm = LengthMm.of(props.start.distanceTo(props.end));
  }

  static create(props: WallProps): Wall {
    if (!(WALL_KINDS as readonly string[]).includes(props.kind)) {
      throw new DomainError(`Unknown wall kind: ${props.kind}`, "INVALID_WALL");
    }
    const length = props.start.distanceTo(props.end);
    if (length <= 0) {
      throw new DomainError("Wall length must be positive", "INVALID_WALL");
    }
    if (props.thickness.value <= 0) {
      throw new DomainError("Wall thickness must be positive", "INVALID_WALL");
    }
    if (props.height.value <= 0) {
      throw new DomainError("Wall height must be positive", "INVALID_WALL");
    }
    const openings = props.openings ?? [];
    const wall = new Wall(props, openings);
    for (const opening of openings) {
      wall.assertOpeningFits(opening);
    }
    return wall;
  }

  hostOpening(opening: Opening): Wall {
    this.assertOpeningFits(opening);
    if (this.openings.some((existing) => existing.id === opening.id)) {
      throw new DomainError(
        `Opening ${opening.id} already hosted on wall ${this.id}`,
        "DUPLICATE_OPENING",
      );
    }
    return new Wall(this, [...this.openings, opening]);
  }

  replaceOpening(opening: Opening): Wall {
    const index = this.openings.findIndex((existing) => existing.id === opening.id);
    if (index < 0) {
      throw new DomainError(
        `Opening ${opening.id} is not hosted on wall ${this.id}`,
        "OPENING_NOT_FOUND",
      );
    }
    this.assertOpeningFits(opening);
    const next = [...this.openings];
    next[index] = opening;
    return new Wall(this, next);
  }

  private assertOpeningFits(opening: Opening): void {
    if (opening.offsetAlongWall.value < 0) {
      throw new DomainError(
        "Opening offset must not be negative",
        "OPENING_OUT_OF_BOUNDS",
      );
    }
    if (opening.occupiesUntilMm() > this.lengthMm.value + 1e-6) {
      throw new DomainError(
        "Opening extends past the host wall length",
        "OPENING_OUT_OF_BOUNDS",
      );
    }
    if (opening.sillHeight.value + opening.height.value > this.height.value + 1e-6) {
      throw new DomainError(
        "Opening is taller than the host wall",
        "OPENING_OUT_OF_BOUNDS",
      );
    }
  }
}

File: core/src/domain/wall.cpp
#include "toporoom/domain/wall.hpp"

#include "toporoom/domain/error.hpp"

namespace toporoom::domain {

Wall::Wall(WallProps props, std::vector<Opening> openings)
    : id_(std::move(props.id)),
      start_(props.start),
      end_(props.end),
      thickness_(props.thickness),
      height_(props.height),
      kind_(props.kind),
      openings_(std::move(openings)),
      length_mm_(LengthMm::of(props.start.distance_to(props.end))) {}

WallProps Wall::snapshot() const {
  WallProps props;
  props.id = id_;
  props.start = start_;
  props.end = end_;
  props.thickness = thickness_;
  props.height = height_;
  props.kind = kind_;
  props.openings = openings_;
  return props;
}

Wall Wall::create(WallProps props) {
  const double length = props.start.distance_to(props.end);
  if (length <= 0) {
    throw DomainError("Wall length must be positive", "INVALID_WALL");
  }
  if (props.thickness.value() <= 0) {
    throw DomainError("Wall thickness must be positive", "INVALID_WALL");
  }
  if (props.height.value() <= 0) {
    throw DomainError("Wall height must be positive", "INVALID_WALL");
  }
  auto openings = std::move(props.openings);
  Wall wall(std::move(props), openings);
  for (const auto& opening : wall.openings_) {
    wall.assert_opening_fits(opening);
  }
  return wall;
}

Wall Wall::host_opening(const Opening& opening) const {
  assert_opening_fits(opening);
  for (const auto& existing : openings_) {
    if (existing.id() == opening.id()) {
      throw DomainError("Opening " + opening.id() + " already hosted on wall " + id_,
                        "DUPLICATE_OPENING");
    }
  }
  auto props = snapshot();
  props.openings.push_back(opening);
  auto openings = props.openings;
  return Wall(std::move(props), std::move(openings));
}

Wall Wall::replace_opening(const Opening& opening) const {
  auto props = snapshot();
  bool found = false;
  for (auto& candidate : props.openings) {
    if (candidate.id() == opening.id()) {
      candidate = opening;
      found = true;
      break;
    }
  }
  if (!found) {
    throw DomainError("Opening " + opening.id() + " is not hosted on wall " + id_,
                      "OPENING_NOT_FOUND");
  }
  return Wall::create(std::move(props));
}

Wall Wall::without_opening(const std::string& opening_id) const {
  auto props = snapshot();
  std::vector<Opening> kept;
  kept.reserve(props.openings.size());
  bool found = false;
  for (const auto& opening : props.openings) {
    if (opening.id() == opening_id) {
      found = true;
      continue;
    }
    kept.push_back(opening);
  }
  if (!found) {
    throw DomainError("Opening " + opening_id + " is not hosted on wall " + id_,
                      "OPENING_NOT_FOUND");
  }
  props.openings = std::move(kept);
  return Wall::create(std::move(props));
}

Wall Wall::with_height(LengthMm height) const {
  auto props = snapshot();
  props.height = height;
  return Wall::create(std::move(props));
}

Wall Wall::with_geometry(PointMm start, PointMm end) const {
  auto props = snapshot();
  props.start = start;
  props.end = end;
  return Wall::create(std::move(props));
}

bool Wall::same_as(PointMm start, PointMm end, LengthMm thickness, LengthMm height,
                   WallKind kind) const noexcept {
  return start_.equals(start) && end_.equals(end) &&
         thickness_.value() == thickness.value() && height_.value() == height.value() &&
         kind_ == kind;
}

void Wall::assert_opening_fits(const Opening& opening) const {
  if (opening.offset_along_wall().value() < 0) {
    throw DomainError("Opening offset must not be negative", "OPENING_OUT_OF_BOUNDS");
  }
  if (opening.occupies_until_mm() > length_mm_.value() + 1e-6) {
    throw DomainError("Opening extends past the host wall length",
                      "OPENING_OUT_OF_BOUNDS");
  }
  if (opening.sill_height().value() + opening.height().value() >
      height_.value() + 1e-6) {
    throw DomainError("Opening is taller than the host wall",
                      "OPENING_OUT_OF_BOUNDS");
  }
}

}  // namespace toporoom::domain
